src/nimblepkg/nimscriptwrapper.nim

```diff
@@ -0,0 +1,131 @@
+# Copyright (C) Andreas Rumpf. All rights reserved.
+# BSD License. Look at license.txt for more info.
+
+## Implements the new configuration system for Nimble. Uses Nim as a
+## scripting language.
+
+import common, version, options, packageinfo, cli
+import hashes, json, os, streams, strutils, strtabs, tables, times, osproc, sets, pegs
+
+type
+  Flags = TableRef[string, seq[string]]
+  ExecutionResult*[T] = object
+    success*: bool
+    command*: string
+    arguments*: seq[string]
+    flags*: Flags
+    retVal*: T
+
+const
+  internalCmd = "e"
+  nimscriptApi = staticRead("nimscriptapi.nim")
+  nimscriptHash = $nimscriptApi.hash().abs()
+
+proc execNimscript(nimsFile, actionName: string, options: Options, live = true): tuple[output: string, exitCode: int] =
+  var
+    cmd = ("nim e --hints:off --verbosity:0 " & nimsFile.quoteShell & " " & actionName).strip()
+
+  if live:
+    result.exitCode = execCmd(cmd)
+    let
+      outFile = nimsFile & ".out"
+    if outFile.fileExists():
+      result.output = outFile.readFile()
+      discard outFile.tryRemoveFile()
+  else:
+    result = execCmdEx(cmd, options = {poUsePath})
+
+proc setupNimscript*(scriptName: string, options: Options): tuple[nimsFile, iniFile: string] =
+  let
+    cacheDir = getTempDir() / "nimblecache"
+    shash = $(scriptName & nimscriptHash).hash().abs()
+    prjCacheDir = cacheDir / scriptName.splitFile().name & "_" & shash
+    nimsCacheFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".nims"
+
+  result.nimsFile = scriptName.parentDir() / scriptName.splitFile().name & "_" & shash & ".nims"
+  result.iniFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".ini"
+
+  if not prjCacheDir.dirExists() or not nimsCacheFile.fileExists() or not result.iniFile.fileExists() or
+    scriptName.getLastModificationTime() > nimsCacheFile.getLastModificationTime():
+    createDir(prjCacheDir)
+    writeFile(nimsCacheFile, nimscriptApi & scriptName.readFile() & "\nonExit()\n")
+    discard tryRemoveFile(result.iniFile)
+
+  if not result.nimsFile.fileExists():
```

Comment:
huh? Why would this not exist? I'm assuming that `result.nimsFile` is the nims file that is being executed, i.e. the .nimble file in my package.

Reply:
The nims file gets generated once and saved in nimblecache and updated if the nimble file changes. When tasks need to be executed, it gets copied into the folder next to the nimble file and executed there. This is because all existing nimble scripts assume they will be in that location, either to import files or access relative paths.

However, we cannot leave the file sitting there since it is unexpected and will stand out with git status, etc. So it is deleted at the end of each call.

Reply:
Right. I would prefer if a single function dealth with this temporary `nims` file, instead of returning it and having another function delete it. I think with the suggestion I gave above this would be possible and make the code more contained.

Reply:
The nims file copy and delete is done in the same proc now.